components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java

```diff
@@ -354,26 +380,56 @@ private static EndpointListRouteDTO extractEndpointFromOpenAPI(Object endpointEx
      *
      * @param openAPI         {@link OpenAPI} object
      * @param openApiFilePath OpenAPI definition file
+     * @param openAPIVersion OpenAPI version
      */
-    private static void validateBasepath(OpenAPI openAPI, String openApiFilePath) {
+    private static void validateBasePath(OpenAPI openAPI, String openApiFilePath, String openAPIVersion) {
+        List<Server> servers = openAPI.getServers();
         Map<String, Object> openAPIExtensions = openAPI.getExtensions();
-        if (openAPIExtensions == null) {
-            throw new CLIRuntimeException(" At least '" + OpenAPIConstants.BASEPATH + "' property and '"
-                    + OpenAPIConstants.PRODUCTION_ENDPOINTS + "' property should present in the open API "
-                    + "definition: '" + openApiFilePath + "'.");
+        String basePath = null;
+        if (openAPIExtensions != null && openAPIExtensions.get(OpenAPIConstants.BASEPATH) != null) {
+            basePath = (String) openAPIExtensions.get(OpenAPIConstants.BASEPATH);
+        } else if (servers != null) {
+            //We  need to check only the first server url to derive the base path. Each and every server url
+            // should contains the same base path.
+            Server server = servers.get(0);
+            if (server.getUrl() != null) {
+                String url = replaceOpenAPIServerTemplate(server);
+                try {
+                    basePath = new URI(url).getPath();
+                } catch (URISyntaxException e) {
+                    String message;
+                    if (openAPISpec2.equals(openAPIVersion)) {
+                        message = "Scheme, host, base path combination of openAPI '" + openApiFilePath
+                                + "' resolves to : " + url + ", which is a malformed url";
+                    } else {
+                        message = "Server url: " + url + " of openAPI '" + openApiFilePath + "' is a malformed url";
+                    }
+                    throw new CLIRuntimeException(message);
+                }
+            }
         }
-        String basePath = (String) openAPIExtensions.get(OpenAPIConstants.BASEPATH);
+
         if (basePath == null || basePath.isEmpty()) {
-            throw new CLIRuntimeException("'" + OpenAPIConstants.BASEPATH + "' property is not included in openAPI " +
-                    "definition '" + openApiFilePath + "'.");
-        }
-        basePath = basePath.startsWith("/") ? basePath : "/" + basePath;
-        if (basePathMap.containsKey(basePath)) {
-            throw new CLIRuntimeException("The value for '" + OpenAPIConstants.BASEPATH + "' " + basePath +
-                    " property is duplicated in the following openAPI definitions.\n" + basePathMap.get(basePath) +
-                    "\n" + openApiFilePath);
+            basePath = "/";
```

Comment:
Having two apis with the same basepath may cause problems in the generated ballerina code. Needs to check that.

Reply:
We maintains the base path map for all the APIS, so befora assigning the base path we check previously added open APIs contains the same base path. This check will fail this during the build time, saying same base path exists in another API also.

Reply:
Let's assume a scenario where two openAPI definitions are provided and both do not have the basepath variable. Then both of them will have the basePath "/". Although the basepath is not mandatory according to this PR, we have to provide the error that the same basepath exists.

Reply:
Yes, there is a warning message is printed on the cases when there is no base path presents and we take default value as "/". The idea is to get build the microgateway without extensions. So the basic scenarios should work without extensions. When the requirements get complicated anyway user will have to use the extensions